### PR TITLE
#3836 - Coalescing string concatenation on nullable Expressions

### PR DESCRIPTION
**Pull request by svengeance**
Fixes https://github.com/aspnet/EntityFrameworkCore/issues/3836

I was hoping to get a review of the auto-coalescing implementation before I fixed the broken tests (if there's a better way than creating a PR with knowingly broken tests let me know (or if I should have fixed the tests first)). This feature is a slight breaking change if anyone was dependent on return nullability, and it also breaks around 30 or so existing tests due to the difference in output SQL.

Based on comments from @smitpatel on the issue we should be able to handle 3 expressions (column/const/param) in the NullSemanticsVisitor and appropriately coalesce these so we can better emulate expected C# functionality

### CODE REVIEW

**Comment by smitpatel** [comment]
@svengeance - Changes look correct.

**Comment by svengeance** [comment]
Good to hear, I'll work on ensuring all broken tests are strictly the result of coalescing during concatenation, fix/commit, and mark the PR as Ready

**Comment by svengeance** [comment]
@smitpatel Looks like I got a CI request timeout? How should I go about requesting a rebuild to validate these checks, or is that out of my hands?

**Comment by smitpatel** [comment]
@svengeance - I kicked off run again.

File: src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs [hunk 1]
```diff
@@ -367,6 +367,30 @@ protected override Expression VisitSqlBinary(SqlBinaryExpression sqlBinaryExpres
                 return sqlBinaryExpression.Update(newLeft, newRight);
             }
 
+            if (sqlBinaryExpression.OperatorType == ExpressionType.Add
+                && sqlBinaryExpression.Type == typeof(string))
+            {
+                if (leftNullable)
+                {
+                    newLeft = newLeft is SqlConstantExpression
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/NullSemanticsRewritingExpressionVisitor.cs]: @maumar - Introduce local function for this.

**Comment by maumar** [comment]
@svengeance thank you for the contribution!

**Pull request merged**